core/src/test/scala/tech/beshu/ror/unit/acl/blocks/definitions/UnboundidLdapAuthenticationServiceTests.scala

```diff
@@ -16,53 +16,92 @@
 package tech.beshu.ror.unit.acl.blocks.definitions
 
-import eu.timepit.refined.auto._
 import cats.data._
 import com.dimafeng.testcontainers.{Container, ForAllTestContainer, MultipleContainers}
+import com.unboundid.ldap.sdk.LDAPSearchException
 import eu.timepit.refined.api.Refined
 import eu.timepit.refined.auto._
 import monix.eval.Task
 import monix.execution.Scheduler.Implicits.global
+import monix.execution.exceptions.ExecutionRejectedException
 import org.scalatest.matchers.should.Matchers._
 import org.scalatest.wordspec.AnyWordSpec
-import org.scalatest.{BeforeAndAfterAll, Inside}
-
-import tech.beshu.ror.accesscontrol.blocks.definitions.ldap.Dn
+import org.scalatest.{BeforeAndAfterAll, BeforeAndAfterEach, Inside}
+import tech.beshu.ror.accesscontrol.blocks.definitions.CircuitBreakerConfig
 import tech.beshu.ror.accesscontrol.blocks.definitions.ldap.LdapService.Name
-import tech.beshu.ror.accesscontrol.blocks.definitions.ldap.implementations.LdapConnectionConfig.{BindRequestUser, ConnectionMethod, HaMethod, LdapHost}
+import tech.beshu.ror.accesscontrol.blocks.definitions.ldap.implementations.LdapConnectionConfig._
 import tech.beshu.ror.accesscontrol.blocks.definitions.ldap.implementations._
+import tech.beshu.ror.accesscontrol.blocks.definitions.ldap.{CircuitBreakerLdapAuthenticationServiceDecorator, Dn, LdapAuthenticationService}
 import tech.beshu.ror.accesscontrol.domain.{PlainTextSecret, User}
 import tech.beshu.ror.utils.ScalaOps.repeat
-import tech.beshu.ror.utils.TestsUtils._
-import tech.beshu.ror.utils.containers.LdapContainer
+import tech.beshu.ror.utils.containers.{LdapContainer, ToxiproxyContainer}
 
 import scala.concurrent.duration._
 import scala.language.postfixOps
+import scala.reflect.ClassTag
 
 class UnboundidLdapAuthenticationServiceTests
   extends AnyWordSpec
     with BeforeAndAfterAll
+    with BeforeAndAfterEach
     with ForAllTestContainer
     with Inside {
 
-  private val ldap1Container = new LdapContainer("LDAP1", "test_example.ldif")
+  private val ldap1Container = LdapContainer.create("LDAP1", "test_example.ldif")
+  private val ldap1ContainerWithToxiproxy = new ToxiproxyContainer(ldap1Container, LdapContainer.defaults.ldap.port)
   private val ldap2Container = new LdapContainer("LDAP2", "test_example.ldif")
-  override val container: Container = MultipleContainers(ldap1Container, ldap2Container)
+  override val container: Container = MultipleContainers(ldap1ContainerWithToxiproxy, ldap2Container)
   private val ldapConnectionPoolProvider = new UnboundidLdapConnectionPoolProvider
 
   override protected def afterAll(): Unit = {
     super.afterAll()
     ldapConnectionPoolProvider.close()
       .runSyncUnsafe()
   }
+
+  override protected def beforeEach(): Unit = {
+    super.beforeEach()
+    ldap1ContainerWithToxiproxy.enableNetwork()
+    ldap1ContainerWithToxiproxy.disableNetworkTimeout()
+  }
+
+  implicit class LdapAuthenticationServiceOps(authenticationService: LdapAuthenticationService) {
```

Comment:
Let's move it at the end of the file (or at least after all test cases - because IMO they are more important than the assertion implmentation)

_(We can do it in a second part of the task)_